### PR TITLE
GCS Bundle Notifier

### PR DESCRIPTION
**Pull request by azdagron**
Stores the latest SPIRE bundle into an object in Google Cloud Storage, similar to the K8S Bundle Notifier.

### CODE REVIEW

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+func (p *Plugin) setConfig(config *pluginConfig) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.config = config
+}
+
+func (p *Plugin) updateBundleObject(ctx context.Context, c *pluginConfig) (err error) {
+	client, err := p.hooks.newBucketClient(ctx, c.ServiceAccountFile)
+	if err != nil {
+		return status.Errorf(codes.Unknown, "unable to instantiate bucket client: %v", err)
```
Comment (evan2645) [inline pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: What conditions can cause this error? Bad service account file?

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	defer client.Close()
+
+	for {
+		// Get the config map so we can use the version to resolve conflicts racing
+		// on updates from other servers.
+		generation, err := client.GetObjectGeneration(ctx, c.Bucket, c.ObjectPath)
+		if err != nil {
+			return status.Errorf(codes.Unknown, "unable to get bundle object %s/%s generation: %v", c.Bucket, c.ObjectPath, err)
+		}
+		p.log.Debug("Bundle object generation retrieved", telemetry.Generation, generation)
```
Comment (evan2645) [inline pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: Perhaps s/Bundle object generation retrieved/Bundle object retrieved/ ? Or "Bundle retrieved from GCS"? Do we expect all GCS/GCP users to know what a "generation" is in this context?

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+		// Load bundle data from the the identity provider. The bundle has to
+		// be loaded after fetching the config map so we can properly detect
```
Comment (evan2645) [inline pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: stale comment

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		// and correct a race updating the bundle (i.e. read-modify-write
+		// semantics).
+		resp, err := p.identityProvider.FetchX509Identity(ctx, &hostservices.FetchX509IdentityRequest{})
+		if err != nil {
+			st := status.Convert(err)
+			return status.Errorf(st.Code(), "unable to fetch bundle from identity provider: %v", st.Message())
```
Comment (evan2645) [inline pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: nit: this message feels like it could be a little friendlier if it referenced spire-server instead of identity provider because knowledge of the plugin implementation is needed in order to know what an/the identity provider is in this context

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+		}
+
+		// Upload the bundle, handling version conflicts
+		if err := client.PutObject(ctx, c.Bucket, c.ObjectPath, bundleData(resp.Bundle), generation); err != nil {
+			// If there is a conflict then some other server won the race updating
+			// the object. We need to retrieve the latest bundle and try again.
+			if e, ok := err.(*googleapi.Error); ok && e.Code == http.StatusPreconditionFailed {
+				p.log.Debug("Conflict detected setting bundle object", telemetry.Generation, generation)
```
Comment (evan2645) [inline pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: nit: might be useful to mention that we will retry here

Comment (evan2645) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: Is it possible for a non-conflict-related problem to cause precondition failed?

File: doc/plugin_server_notifier_gcpbundle.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
+# Server plugin: Notifier "gcpbundle"
```
[the remark this answers is not in the document]
Comment (azdagron) [reply doc/plugin_server_notifier_gcpbundle.md]: 👍 i copied most of this from `k8sbundle`..... which should have been `k8sbundle` i guess?

File: doc/plugin_server_notifier_gcpbundle.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+The `gcpbundle` plugin responds to bundle loaded/updated events by fetching and
+pushing the latest root CA certificates from the trust bundle to an object in
+Google Cloud Storage.
+
+The certificates in the object can be used to bootstrap SPIRE agents.
+
+In HA, multiple SPIRE servers will be pushing bundle contents to the object.
+The plugin handles conflict resolution by implementing a read-modify-write
+cycle to ensure the object contains the latest bundle contents.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply doc/plugin_server_notifier_gcpbundle.md]: 👍

File: doc/plugin_server_notifier_gcpbundle.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+| ---------------------- | ------------------------------------------- | --------------- |
+| `bucket`               | The bucket containing the object            |                 |
+| `object_path`          | The path to the object within the bucket    |                 |
+| `service_account_file` | Path to the service account file used to authenticate with Google Cloud Storage | |
+
+## Sample configurations
+
+### Default In-Cluster
+
+The following configuration pushes bundle contents from an in-cluster SPIRE
```
[the remark this answers is not in the document]
Comment (azdagron) [reply doc/plugin_server_notifier_gcpbundle.md]: I'll rephrase this to make it more clear.

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go (hunk 1 above)
Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: Practically speaking that is probably the case.

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go (hunk 2 above)
Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: 👍

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go (hunk 3 above)
Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: 👍

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go (hunk 4 above)
Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: 👍

File: pkg/server/plugin/notifier/gcpbundle/gcpbundle.go (hunk 5 above)
Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: Good question. Here are the reasons defined by https://cloud.google.com/storage/docs/json_api/v1/status-codes.

```
conditionNotMet | At least one of the pre-conditions you specified did not hold.
orgPolicyConstraintFailed | Request violates an OrgPolicy constraint.
```

The error type has a list of error times that contain the "reason" field. I can drill down more specifically to make sure `conditionNotMet` is one of the reasons.

Comment (azdagron) [reply pkg/server/plugin/notifier/gcpbundle/gcpbundle.go]: I've tested this against GCS and confirmed the shape of the error returned in this situation (included in the coded from 17f71f5)

File: doc/plugin_server_notifier_gcpbundle.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (evan2645) [reply doc/plugin_server_notifier_gcpbundle.md]: Yea... we've been bad about this in the past, I'm trying to get better about being more consistent
